### PR TITLE
Replace hero image with carousel slider

### PR DESCRIPTION
**Pull request by TimOsahenru**
## Issue Link 🔗:

**Issue**: #365 

## Type of Change

- [ ] Bug fix 🐞
- [x] New feature/page
- [ ] Documentation update
- [ ] Other

## Description 📋

Replaced static hero image with a dynamic Bootstrap carousel for improved user experiece.

## Checklist ✅

- [x] Followed the [Code of Conduct](https://github.com/BlackPythonDevs/blackpythondevs.github.io?tab=coc-ov-file) and [Contribution Guide](https://github.com/BlackPythonDevs/blackpythondevs.github.io/blob/469afd71943738b7209bfe56d5a4e35394fbaafc/CONTRIBUTING.md)
- [x] Ran `pre-commit run --all`
- [x] All tests pass locally
- [x] Added tests (if applicable)
- [ ] Documentation updated (if applicable)

## Additional Notes & Screenshots

Add any additional notes or comments that might be helpful for the reviewers.

### CODE REVIEW

**Comment by kjaymiller** [comment]
@TimOsahenru I'm glad to hear that you were able to get the tests passing? What did you wind up doing in the end?

File: tests/test.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ def page_url(xprocess, url_port):
     url, port = url_port
 
     class Starter(ProcessStarter):
-        timeout = 4
+        timeout = 20
```
Comment (kjaymiller) [inline tests/test.py:15]: Ah so your server wasn't ready yet. This is a good solution... great work troubleshooting and catching this!!

**Comment by kjaymiller** [comment]
@TimOsahenru,

This is a great start!

There are a couple of things that I would like to see refined before we go live with the changes. That said I don't want to lose this work so I've created a new branch called `New-Carousel`... can you submit the PR to that branch and then we can look at some of the things that I would hope that we can get touched up?

Quick List of things to address:

- Only show the blog posts that are in recent posts. This also means that we can de-emphasize that section and make it smaller.
- The circle image makes it hard to see some images, it may be better as a full width image`
- With the new featured image setup, we should be able to use that for articles with
- The descriptions seem like the first sentence of the content and not a dedicated line that can be used for MetaTagging as well. - I recently created a tool that would generate autodescriptions for my personal blog that could be re-used for this. -   https://github.com/kjaymiller/kjaymiller.com/blob/main/ollama_summarizer.py `
- The section is margined inside the container making it smaller than the content near it. We can fix that by changing the width of the container from 85% to 100%
- The default image seems to be a little buggy. Instead of giving the default image it like creates an image of the raw content... we can fix this by removing `default: post.content | split: '!' | first | split: '(' | last | split: ')' | first |`

File: _layouts/default.html [hunk 2]
```diff
@@ -12,6 +12,7 @@
     <link rel="canonical" href="{{ page.url | replace:'index.html','' | prepend: site.baseurl | prepend: site.url }}" />
     <link rel="alternate" type="application/rss+xml" title="{{ site.title }}" href="{{ '/feed.xml' | prepend: site.baseurl | prepend: site.url }}" />
     <link rel="icon" href="https://fav.farm/%E2%9C%8A%F0%9F%8F%BE" type="image/svg" />
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.0-alpha1/dist/css/bootstrap.min.css" rel="stylesheet">
```
Comment (kjaymiller) [inline _layouts/default.html:15]: You chose to use bootstrap. Interesting... I wonder what the load time difference is between using bootstrap versus a solution that uses css with transitions and animations.

Adding bootstrap for a singular component doesn't feel great.

File: _layouts/index.html [hunk 3]
```diff
@@ -3,9 +3,31 @@
...
+  <header>
+    <div id="postSlider" class="carousel slide slider-container" data-bs-ride="carousel">
+      <div class="carousel-inner">
+        {% for post in site.posts %}
+        <div class="carousel-item {% if forloop.first %}active{% endif %}">
+          <div class="email-circle">
+            <img src="{{ post.featured_image | default: post.content | split: '!' | first | split: '(' | last | split: ')' | first | default: 'default-image.jpg' }}" alt="{{ post.title }}">
+          </div>
+          <h3 class="post-title">{{ post.title }}</h3>
+          <p class="post-summary">{{ post.excerpt | strip_html | truncatewords: 20 }}</p>
```
Comment (kjaymiller) [inline _layouts/index.html:15]: This creates some interesting issues... Perhaps we can request a description on each article (I mentioned this in the comments)

File: _layouts/index.html [hunk 4]
```diff
@@ -3,9 +3,31 @@
...
-  <article class="hero">
-    <div class="hero-text">{{ t.index.lead }}</div>
-  </article>
+  <header>
+    <div id="postSlider" class="carousel slide slider-container" data-bs-ride="carousel">
+      <div class="carousel-inner">
+        {% for post in site.posts %}
+        <div class="carousel-item {% if forloop.first %}active{% endif %}">
+          <div class="email-circle">
+            <img src="{{ post.featured_image | default: post.content | split: '!' | first | split: '(' | last | split: ')' | first | default: 'default-image.jpg' }}" alt="{{ post.title }}">
```
Comment (kjaymiller) [inline _layouts/index.html:12]: This is creating some issues in that the default image is not being selected.

File: assets/css/bpd.css [hunk 5]
```diff
@@ -517,3 +517,50 @@ ul.speaking-list {
...
+  min-height: 300px;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  /* background-image: url("/assets/images/hero-bpd-background.jpg"); */
+  background-color: #222;
+  padding: 40px;
+  border-radius: 10px;
+  margin: 0 auto;
+  width: 85%;
```
Comment (kjaymiller) [inline assets/css/bpd.css:535]: This width inside the container makes the carousel smaller than everything around it. I'd suggest making it 100%.

File: assets/css/bpd.css [hunk 6]
```diff
@@ -517,3 +517,50 @@ ul.speaking-list {
...
+  /* max-width: 1000px; */
+}
+.email-circle {
```
Comment (kjaymiller) [inline assets/css/bpd.css:538]: Why is this called email circle... 

Naming is important... I hope we can name it something that matches the description.

File: tests/test.py [hunk 7]
```diff
@@ -99,3 +99,16 @@ def test_mailto_bpdevs(page_url: tuple[Page, str]) -> None:
     page.goto(live_server_url)
     mailto = page.get_by_role("link", name="email")
     expect(mailto).to_have_attribute("href", "mailto:contact@blackpythondevs.com")
+
```
Comment (kjaymiller) [inline tests/test.py:102]: Some tests that we have coded for that I would like to see.

#### Test 1:
- Add a small post as a demo (in test as a test case and destroy it after)
- Wait and let it rebuild and then check to see if the default image is grabbed.
 
 #### Test 2
 - If you select the next_button you expect the image to change to a new one. Same for prev_button
 
 #### Test 3
 - That slides match the links in the recent posts section
 
 #### Test 4
 - When you click the links in the carousel they go to the corresponding webpages

**Review by kjaymiller — CHANGES_REQUESTED** [review]
@TimOsahenru great work on this...

I know there are a lot of comments which is why I want to use this as the start for the work in `New Carousel` and then iterate on that work to get it live on the page. This has been excellent so far!

**Comment by TimOsahenru** [comment]
> @TimOsahenru I'm glad to hear that you were able to get the tests passing? What did you wind up doing in the end?

I tried a couple of unconventional things 😆, but ultimately increasing the server time from 4 to 20 was what did the trick for me.

**Comment by TimOsahenru** [comment]
> @TimOsahenru,
> 
> This is a great start!
> 
> There are a couple of things that I would like to see refined before we go live with the changes. That said I don't want to lose this work so I've created a new branch called `New-Carousel`... can you submit the PR to that branch and then we can look at some of the things that I would hope that we can get touched up?
> 
> Quick List of things to address:
> 
> * Only show the blog posts that are in recent posts. This also means that we can de-emphasize that section and make it smaller.
> * The circle image makes it hard to see some images, it may be better as a full width image`
> * With the new featured image setup, we should be able to use that for articles with
> * The descriptions seem like the first sentence of the content and not a dedicated line that can be used for MetaTagging as well. - I recently created a tool that would generate autodescriptions for my personal blog that could be re-used for this. -   https://github.com/kjaymiller/kjaymiller.com/blob/main/ollama_summarizer.py `
> * The section is margined inside the container making it smaller than the content near it. We can fix that by changing the width of the container from 85% to 100%
> * The default image seems to be a little buggy. Instead of giving the default image it like creates an image of the raw content... we can fix this by removing `default: post.content | split: '!' | first | split: '(' | last | split: ')' | first |`

Thanks. I'll  effect all these changes and make a push to `New-Carousel` branch

File: _layouts/default.html (hunk 2 above)
Comment (TimOsahenru) [reply _layouts/default.html:15]: You’re absolutely right; I hadn’t fully considered the impact of load time when choosing Bootstrap. I felt custom CSS might be re-inventing the wheel.

I’ll look into using a custom CSS and explore the impact both has on our load time. Thanks for the insight

File: _layouts/index.html (hunk 4 above)
Comment (TimOsahenru) [reply _layouts/index.html:12]: Gotcha! Working on it. Thanks

File: tests/test.py (hunk 7 above)
Comment (TimOsahenru) [reply tests/test.py:102]: Thanks Jay. I'll get to implementing these changes right away

**Pull request merged**